File: ios/flutter_comics_viewer/Sources/flutter_comics_viewer/ComicsViewerPlatformView.swift
import Flutter
import UIKit
import ComicsViewer

class ComicsViewerPlatformView: NSObject, FlutterPlatformView {
    private let _view: UIView
    private let methodChannel: FlutterMethodChannel
    private let imageScrollView: ImageScrollView
    private var controller: ComicsViewerController!

    private var filePath: String?
    private var languageIndex: Int = 0
    private var soundEnabled: Bool = true

    init(
        frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?,
        binaryMessenger messenger: FlutterBinaryMessenger
    ) {
        _view = UIView(frame: frame)
        _view.backgroundColor = .black

        methodChannel = FlutterMethodChannel(
            name: "flutter_comics_viewer_\(viewId)",
            binaryMessenger: messenger
        )

        // Initialize ImageScrollView from ComicsViewer package
        imageScrollView = ImageScrollView()
        imageScrollView.isComics = true
        imageScrollView.translatesAutoresizingMaskIntoConstraints = false

        super.init()

        // Extract creation parameters
        if let arguments = args as? [String: Any] {
            filePath = arguments["filePath"] as? String
            languageIndex = arguments["languageIndex"] as? Int ?? 0
            soundEnabled = arguments["soundEnabled"] as? Bool ?? true
        }

        imageScrollView.languageIndex = languageIndex
        imageScrollView.soundEnabled = soundEnabled

        _view.addSubview(imageScrollView)
        NSLayoutConstraint.activate([
            imageScrollView.topAnchor.constraint(equalTo: _view.topAnchor),
            imageScrollView.leadingAnchor.constraint(equalTo: _view.leadingAnchor),
            imageScrollView.trailingAnchor.constraint(equalTo: _view.trailingAnchor),
            imageScrollView.bottomAnchor.constraint(equalTo: _view.bottomAnchor)
        ])

        // Initialize controller
        controller = ComicsViewerController(scrollView: imageScrollView)

        // Setup scroll listener
        controller.onScrollChanged = { [weak self] position in
            self?.methodChannel.invokeMethod("onScrollChanged", arguments: ["position": position])
        }

        // Load comics if file path is provided
        if let path = filePath {
            loadComics(path: path)
        }

        methodChannel.setMethodCallHandler { [weak self] (call, result) in
            self?.handleMethodCall(call, result: result)
        }
    }

    func view() -> UIView {
        return _view
    }

    private func handleMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadComics":
            guard let args = call.arguments as? [String: Any],
                  let path = args["filePath"] as? String else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "File path is required", details: nil))
                return
            }
            loadComics(path: path)
            result(nil)

        case "play":
            controller.play()
            result(nil)

        case "pause":
            controller.pause()
            result(nil)

        case "setScrollPosition":
            guard let args = call.arguments as? [String: Any],
                  let position = args["position"] as? Double else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Position is required", details: nil))
                return
            }
            controller.setScrollPosition(CGFloat(position))
            result(nil)

        case "getScrollPosition":
            let position = controller.getScrollPosition()
            result(Double(position))

        case "togglePreview":
            guard let args = call.arguments as? [String: Any],
                  let show = args["show"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Show flag is required", details: nil))
                return
            }
            controller.togglePreview(show)
            result(nil)

        case "toggleSounds":
            guard let args = call.arguments as? [String: Any],
                  let enabled = args["enabled"] as? Bool else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Enabled flag is required", details: nil))
                return
            }
            controller.toggleSounds(enabled)
            soundEnabled = enabled
            result(nil)

        case "setLanguage":
            guard let args = call.arguments as? [String: Any],
                  let index = args["languageIndex"] as? Int else {
                result(FlutterError(code: "INVALID_ARGUMENT", message: "Language index is required", details: nil))
                return
            }
            controller.setLanguage(index)
            languageIndex = index
            result(nil)

        case "isPlaying":
            result(controller.isPlaying)

        case "getDuration":
            result(Double(controller.duration))

        case "getCurrentPosition":
            result(Double(controller.currentPosition))

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func loadComics(path: String) {
        controller.loadComics(filePath: path) { [weak self] result in
            switch result {
            case .success:
                self?.methodChannel.invokeMethod("onLoaded", arguments: nil)
            case .failure(let error):
                self?.methodChannel.invokeMethod("onError", arguments: ["error": error.localizedDescription])
            }
        }
    }

    deinit {
        controller.dispose()
        methodChannel.setMethodCallHandler(nil)
    }
}

File: android/src/main/kotlin/net/nativemind/flutter/comics/viewer/ComicsViewerPlatformView.kt
package net.nativemind.flutter.comics.viewer

import android.content.Context
import android.view.View
import android.widget.FrameLayout
import android.widget.ScrollView
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView

import net.nativemind.comics.viewer.ComicsViewController
import net.nativemind.comics.viewer.comics.view.LayersView

/**
 * Platform view that displays the comics viewer
 */
class ComicsViewerPlatformView(
    private val context: Context,
    id: Int,
    creationParams: Map<String, Any>?,
    messenger: BinaryMessenger
) : PlatformView, MethodChannel.MethodCallHandler {

    private val containerView: FrameLayout = FrameLayout(context)
    private val methodChannel: MethodChannel = MethodChannel(messenger, "flutter_comics_viewer_$id")

    private val scrollView: ScrollView = ScrollView(context)
    private lateinit var layersView: LayersView
    private lateinit var controller: ComicsViewController

    private var filePath: String? = null
    private var languageIndex: Int = 0
    private var soundEnabled: Boolean = true
    private var muted: Boolean = false

    init {
        methodChannel.setMethodCallHandler(this)

        // Extract creation parameters
        filePath = creationParams?.get("filePath") as? String
        languageIndex = (creationParams?.get("languageIndex") as? Int) ?: 0
        soundEnabled = (creationParams?.get("soundEnabled") as? Boolean) ?: true

        // Initialize views
        layersView = LayersView(context, null)

        scrollView.addView(
            layersView,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.WRAP_CONTENT
            )
        )

        containerView.addView(
            scrollView,
            FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
        )

        // Initialize controller
        controller = ComicsViewController(context, layersView)

        // Setup scroll listener
        controller.setScrollListener { position ->
            methodChannel.invokeMethod("onScrollChanged", mapOf("position" to position))
        }

        // Load comics if file path is provided
        filePath?.let { path ->
            loadComics(path)
        }
    }

    override fun getView(): View = containerView

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "loadComics" -> {
                val path = call.argument<String>("filePath")
                val requestId = call.argument<Int>("requestId")
                if (path != null) {
                    loadComics(path, requestId)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "File path is required", null)
                }
            }
            "play" -> {
                controller.play()
                result.success(null)
            }
            "pause" -> {
                controller.pause()
                result.success(null)
            }
            "setScrollPosition" -> {
                val position = call.argument<Double>("position")
                if (position != null) {
                    controller.setScrollPosition(position.toFloat())
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Position is required", null)
                }
            }
            "getScrollPosition" -> {
                val position = controller.getScrollPosition()
                result.success(position.toDouble())
            }
            "togglePreview" -> {
                val show = call.argument<Boolean>("show")
                if (show != null) {
                    controller.togglePreview(show)
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Show flag is required", null)
                }
            }
            "setSoundEnabled" -> {
                val enabled = call.argument<Boolean>("enabled")
                if (enabled != null) {
                    soundEnabled = enabled
                    applySoundState()
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Enabled flag is required", null)
                }
            }
            "setMuted" -> {
                val value = call.argument<Boolean>("muted")
                if (value != null) {
                    muted = value
                    applySoundState()
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Muted flag is required", null)
                }
            }
            "setLanguageIndex" -> {
                val index = call.argument<Int>("index")
                if (index != null) {
                    controller.setLanguage(index)
                    languageIndex = index
                    result.success(null)
                } else {
                    result.error("INVALID_ARGUMENT", "Language index is required", null)
                }
            }
            "isPlaying" -> {
                result.success(controller.isPlaying())
            }
            "getDuration" -> {
                result.success(controller.getDuration().toDouble())
            }
            "getCurrentPosition" -> {
                result.success(controller.getCurrentPosition().toDouble())
            }
            else -> result.notImplemented()
        }
    }

    private fun applySoundState() {
        controller.toggleSounds(soundEnabled && !muted)
    }

    private fun loadComics(path: String, requestId: Int? = null) {
        controller.loadComics(path, object : ComicsViewController.ComicsLoadListener {
            override fun onLoaded() {
                methodChannel.invokeMethod("onLoaded", mapOf("requestId" to requestId))
            }

            override fun onError(error: String) {
                methodChannel.invokeMethod(
                    "onError",
                    mapOf("error" to error, "requestId" to requestId)
                )
            }
        })
    }

    override fun dispose() {
        controller.dispose()
        methodChannel.setMethodCallHandler(null)
    }
}
